File: upload_images.mjs
/**
 * Sube las imágenes optimizadas a Shopify CDN y actualiza los productos.
 * Uso: node upload_images.mjs
 */

import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

import fs from "fs";
let TOKEN = "";
try {
  const envFile = fs.readFileSync(".env", "utf8");
  const match = envFile.match(/SHOPIFY_ACCESS_TOKEN=(.*)/);
  if (match) TOKEN = match[1].trim();
} catch (e) {}
const SHOP  = "mueblesexterior.myshopify.com";
const API   = `https://${SHOP}/admin/api/2026-01/graphql.json`;
const BASE  = path.dirname(fileURLToPath(import.meta.url));

// ── GraphQL helper ────────────────────────────────────────────────────────────

async function gql(query, variables = {}) {
  const res = await fetch(API, {
    method:  "POST",
    headers: { "X-Shopify-Access-Token": TOKEN, "Content-Type": "application/json" },
    body:    JSON.stringify({ query, variables }),
  });
  const json = await res.json();
  if (json.errors) throw new Error(JSON.stringify(json.errors));
  return json.data;
}

// ── Staged upload ─────────────────────────────────────────────────────────────

async function stagedUpload(filename, mimeType, fileSize) {
  const data = await gql(`
    mutation stagedUploadsCreate($input: [StagedUploadInput!]!) {
      stagedUploadsCreate(input: $input) {
        stagedTargets {
          url
          resourceUrl
          parameters { name value }
        }
        userErrors { field message }
      }
    }
  `, {
    input: [{
      filename,
      mimeType,
      fileSize: String(fileSize),
      resource: "IMAGE",
      httpMethod: "POST",
    }],
  });
  const errs = data.stagedUploadsCreate.userErrors;
  if (errs.length) throw new Error(errs.map(e => e.message).join(", "));
  return data.stagedUploadsCreate.stagedTargets[0];
}

async function uploadToS3(target, filePath, mimeType) {
  const fileBuffer = fs.readFileSync(filePath);
  const boundary   = "----FormBoundary" + Math.random().toString(36).slice(2);

  let headerStr = "";
  for (const { name, value } of target.parameters) {
    headerStr += `--${boundary}\r\nContent-Disposition: form-data; name="${name}"\r\n\r\n${value}\r\n`;
  }
  headerStr += `--${boundary}\r\nContent-Disposition: form-data; name="file"; filename="${path.basename(filePath)}"\r\nContent-Type: ${mimeType}\r\n\r\n`;

  const body = Buffer.concat([
    Buffer.from(headerStr),
    fileBuffer,
    Buffer.from(`\r\n--${boundary}--\r\n`),
  ]);

  const uploadRes = await fetch(target.url, {
    method:  "POST",
    headers: { "Content-Type": `multipart/form-data; boundary=${boundary}` },
    body,
  });

  if (!uploadRes.ok) {
    const txt = await uploadRes.text();
    throw new Error(`S3 upload failed ${uploadRes.status}: ${txt.slice(0, 300)}`);
  }
  return target.resourceUrl;
}

// ── Shopify product helpers ───────────────────────────────────────────────────

async function getProductByHandle(handle) {
  const data = await gql(`
    query($handle: String!) {
      productByHandle(handle: $handle) {
        id
        images(first: 20) {
          edges { node { id src } }
        }
      }
    }
  `, { handle });
  return data.productByHandle;
}

async function appendProductImage(productId, src) {
  const data = await gql(`
    mutation productAppendImages($input: ProductAppendImagesInput!) {
      productAppendImages(input: $input) {
        newImages { id src }
        userErrors { field message }
      }
    }
  `, { input: { id: productId, images: [{ src }] } });
  const errs = data.productAppendImages.userErrors;
  if (errs.length) throw new Error(errs.map(e => e.message).join(", "));
  return data.productAppendImages.newImages[0];
}

// ── CSV parser ────────────────────────────────────────────────────────────────

function parseCSV(filePath) {
  const lines = fs.readFileSync(filePath, "utf-8").split("\n");
  const headers = parseCSVLine(lines[0]);
  return lines.slice(1).filter(l => l.trim()).map(l => {
    const vals = parseCSVLine(l);
    return Object.fromEntries(headers.map((h, i) => [h, vals[i] ?? ""]));
  });
}

function parseCSVLine(line) {
  const result = [];
  let current = "";
  let inQuotes = false;
  for (let i = 0; i < line.length; i++) {
    const ch = line[i];
    if (ch === '"') {
      if (inQuotes && line[i + 1] === '"') { current += '"'; i++; }
      else inQuotes = !inQuotes;
    } else if (ch === "," && !inQuotes) {
      result.push(current); current = "";
    } else {
      current += ch;
    }
  }
  result.push(current);
  return result;
}

// ── main ──────────────────────────────────────────────────────────────────────

async function main() {
  const rows = parseCSV(path.join(BASE, "shopify_products_optimized.csv"));

  // Collect unique localPath → [handles]
  const imageMap = new Map();
  for (const row of rows) {
    const img = (row["Image Src"] || "").trim();
    if (!img.startsWith("images_optimized/")) continue;
    const localPath = path.join(BASE, img);
    if (!fs.existsSync(localPath)) {
      console.warn(`⚠ No existe: ${img}`);
      continue;
    }
    if (!imageMap.has(img)) imageMap.set(img, []);
    const handle = row["Handle"];
    if (handle && !imageMap.get(img).includes(handle)) imageMap.get(img).push(handle);
  }

  console.log(`\nImágenes a subir: ${imageMap.size}\n`);

  const cdnMap = new Map();
  let i = 0;

  // Step 1: Upload all images
  for (const [imgPath] of imageMap) {
    i++;
    const localPath = path.join(BASE, imgPath);
    const filename  = path.basename(localPath);
    const fileSize  = fs.statSync(localPath).size;

    process.stdout.write(`[${i}/${imageMap.size}] ${filename.padEnd(50)} `);
    try {
      const target = await stagedUpload(filename, "image/jpeg", fileSize);
      const cdnUrl = await uploadToS3(target, localPath, "image/jpeg");
      cdnMap.set(imgPath, cdnUrl);
      console.log("✓");
    } catch (err) {
      console.log(`✗ ${err.message.slice(0, 80)}`);
    }
  }

  console.log(`\n── Actualizando productos (${cdnMap.size} imágenes subidas) ────────────\n`);

  // Step 2: Update products
  const done = new Set();
  for (const [imgPath, handles] of imageMap) {
    const cdnUrl = cdnMap.get(imgPath);
    if (!cdnUrl) continue;

    for (const handle of handles) {
      const key = handle + "|" + imgPath;
      if (done.has(key)) continue;
      done.add(key);

      try {
        const product = await getProductByHandle(handle);
        if (!product) { console.log(`  ⚠ No encontrado: ${handle}`); continue; }

        const filename = path.basename(imgPath);
        const alreadyHas = product.images.edges.some(e => e.node.src.includes(filename));
        if (alreadyHas) { console.log(`  ✓ ${handle} (ya tiene imagen)`); continue; }

        await appendProductImage(product.id, cdnUrl);
        console.log(`  ✓ ${handle}`);
      } catch (err) {
        console.log(`  ✗ ${handle}: ${err.message.slice(0, 100)}`);
      }
    }
  }

  console.log("\n✅ Completado.\n");
}

main().catch(err => { console.error("Fatal:", err); process.exit(1); });

File: upload_images.py
#!/usr/bin/env python3
"""
Sube imágenes optimizadas a Shopify CDN y actualiza los productos.
Uso: python3 upload_images.py
"""

import csv
import json
import os
import sys
import time
import urllib.request
import urllib.error
from pathlib import Path

from config import SHOPIFY_ACCESS_TOKEN as TOKEN
SHOP  = "mueblesexterior.myshopify.com"
API   = f"https://{SHOP}/admin/api/2026-01/graphql.json"
BASE  = Path(__file__).parent

# ── GraphQL helper ────────────────────────────────────────────────────────────

def gql(query, variables=None):
    payload = json.dumps({"query": query, "variables": variables or {}}).encode()
    req = urllib.request.Request(
        API,
        data=payload,
        headers={
            "X-Shopify-Access-Token": TOKEN,
            "Content-Type": "application/json",
        },
        method="POST",
    )
    with urllib.request.urlopen(req, timeout=30) as r:
        data = json.loads(r.read())
    if data.get("errors"):
        raise RuntimeError(json.dumps(data["errors"])[:300])
    return data["data"]

# ── Staged upload ─────────────────────────────────────────────────────────────

def staged_upload(filename, file_size):
    data = gql("""
        mutation stagedUploadsCreate($input: [StagedUploadInput!]!) {
          stagedUploadsCreate(input: $input) {
            stagedTargets {
              url
              resourceUrl
              parameters { name value }
            }
            userErrors { field message }
          }
        }
    """, {"input": [{
        "filename":   filename,
        "mimeType":   "image/jpeg",
        "fileSize":   str(file_size),
        "resource":   "IMAGE",
        "httpMethod": "POST",
    }]})
    errs = data["stagedUploadsCreate"]["userErrors"]
    if errs:
        raise RuntimeError(", ".join(e["message"] for e in errs))
    return data["stagedUploadsCreate"]["stagedTargets"][0]

def upload_to_s3(target, file_path):
    file_bytes = Path(file_path).read_bytes()
    boundary   = b"----FormBoundary7MA4YWxkTrZu0gW"

    body_parts = []
    for p in target["parameters"]:
        body_parts.append(
            b"--" + boundary + b"\r\n"
            b'Content-Disposition: form-data; name="' + p["name"].encode() + b'"\r\n\r\n'
            + p["value"].encode() + b"\r\n"
        )
    body_parts.append(
        b"--" + boundary + b"\r\n"
        b'Content-Disposition: form-data; name="file"; filename="' + Path(file_path).name.encode() + b'"\r\n'
        b"Content-Type: image/jpeg\r\n\r\n"
        + file_bytes + b"\r\n"
    )
    body_parts.append(b"--" + boundary + b"--\r\n")
    body = b"".join(body_parts)

    req = urllib.request.Request(
        target["url"],
        data=body,
        headers={"Content-Type": f"multipart/form-data; boundary={boundary.decode()}"},
        method="POST",
    )
    try:
        with urllib.request.urlopen(req, timeout=60) as r:
            pass
    except urllib.error.HTTPError as e:
        raise RuntimeError(f"S3 {e.code}: {e.read()[:200]}")
    return target["resourceUrl"]

# ── Product helpers ───────────────────────────────────────────────────────────

def get_product_by_handle(handle):
    data = gql("""
        query($handle: String!) {
          productByHandle(handle: $handle) {
            id
            images(first: 20) {
              edges { node { id src } }
            }
          }
        }
    """, {"handle": handle})
    return data.get("productByHandle")

def create_product_media(product_id, cdn_url):
    data = gql("""
        mutation productCreateMedia($productId: ID!, $media: [CreateMediaInput!]!) {
          productCreateMedia(productId: $productId, media: $media) {
            media {
              ... on MediaImage {
                id
                image { url }
              }
            }
            mediaUserErrors { field message }
          }
        }
    """, {
        "productId": product_id,
        "media": [{"originalSource": cdn_url, "mediaContentType": "IMAGE"}],
    })
    errs = data["productCreateMedia"]["mediaUserErrors"]
    if errs:
        raise RuntimeError(", ".join(e["message"] for e in errs))
    return data["productCreateMedia"]["media"]

# ── main ──────────────────────────────────────────────────────────────────────

def main():
    csv_path = BASE / "shopify_products_optimized.csv"
    with open(csv_path, newline="", encoding="utf-8") as f:
        rows = list(csv.DictReader(f))

    # Build map: local_img_path → [handles]
    image_map = {}
    for row in rows:
        img = row.get("Image Src", "").strip()
        if not img.startswith("images_optimized/"):
            continue
        local = BASE / img
        if not local.exists():
            print(f"⚠ No existe: {img}")
            continue
        if img not in image_map:
            image_map[img] = []
        handle = row.get("Handle", "").strip()
        if handle and handle not in image_map[img]:
            image_map[img].append(handle)

    print(f"\nImágenes a subir: {len(image_map)}\n")

    cdn_map = {}
    for i, (img_path, handles) in enumerate(image_map.items(), 1):
        local = BASE / img_path
        filename = local.name
        file_size = local.stat().st_size
        label = f"[{i}/{len(image_map)}] {filename[:48]:<48}"
        sys.stdout.write(f"{label} ")
        sys.stdout.flush()
        try:
            target  = staged_upload(filename, file_size)
            cdn_url = upload_to_s3(target, local)
            cdn_map[img_path] = cdn_url
            print("✓")
        except Exception as e:
            print(f"✗  {str(e)[:80]}")

    print(f"\n── Actualizando productos ({len(cdn_map)} imágenes subidas) ────────────\n")

    done = set()
    for img_path, handles in image_map.items():
        cdn_url = cdn_map.get(img_path)
        if not cdn_url:
            continue
        for handle in handles:
            key = (handle, img_path)
            if key in done:
                continue
            done.add(key)
            try:
                product = get_product_by_handle(handle)
                if not product:
                    print(f"  ⚠ No encontrado: {handle}")
                    continue
                filename = Path(img_path).name
                already  = any(filename in e["node"]["src"] for e in product["images"]["edges"])
                if already:
                    print(f"  ✓ {handle[:70]} (ya tiene imagen)")
                    continue
                create_product_media(product["id"], cdn_url)
                print(f"  ✓ {handle[:70]}")
                time.sleep(0.3)  # rate limit
            except Exception as e:
                print(f"  ✗ {handle[:60]}: {str(e)[:100]}")

    print("\n✅ Completado.\n")

if __name__ == "__main__":
    main()
